### PR TITLE
Kubernetes: Added a helm chart to bootstrap skyring to a kubernetes c…

### PR DESCRIPTION
**Pull request by Iaskebba**
…luster

### CODE REVIEW

**Comment by Iaskebba** [comment]
Changed PR To add option for Mongo for persistency. Also I do have a question for @esatterwhite . In this chart I do not seed using 3456. Is there a reason in the examples why you seed on both 3455 and 3456?

**Comment by esatterwhite** [comment]
there is nothing specific about the port numbers themselves. They were just easy to type and didn't conflict when running a cluster locally.

I generally like to start a cluster with more than one seed node to make sure a stable communication ring is established before allowing more nodes to join. It isn't necessarily required though.

File: kubernetes/skyring/charts/mongodb/.helmignore [hunk 1]
```diff
@@ -0,0 +1 @@
+.git
```
Comment (esatterwhite) [inline kubernetes/skyring/charts/mongodb/.helmignore]: You shouldn't have to commit the `charts` directory. 
the `requirements.yml` should be all helm needs.

File: kubernetes/skyring/values.yaml [hunk 2]
```diff
@@ -0,0 +1,35 @@
+# Default values for skyring.
+# This is a YAML-formatted file.
+
+SeedReplicaCount: 2
+WorkerReplicaCount: 4
+
+image:
+  repository: ianskebba/skyring-node
```
Comment (esatterwhite) [inline kubernetes/skyring/values.yaml]: can we use one of the images from [quay](https://quay.io/repository/esatterwhite/skyring)? `7-latest` probably the most recent

**Comment by esatterwhite** [comment]
also, if you are running on separate hosts or in containers / k8s. the ports numbers specifically are a little less important - every node could just use `<HOST>:3456` and be fine

File: kubernetes/skyring/values.yaml (hunk 2 above)
Comment (Iaskebba) [reply kubernetes/skyring/values.yaml]: Is there a version on quay that has the mongo dependency?

**Comment by Iaskebba** [comment]
@esatterwhite I'm trying to add support for scylladb as well due to the design being more resilient to node failures. However when I'm trying to use the scylla configs the skyring nodes don't actually connect to one another. The environmental variables I am using are 
`
DEBUG: '*'
channel__host: 'node-1'
nats__hosts: 'nats-a:4222,nats-b:4222,nats-c:4222'
seeds: 'node-1:3455'
storage__backend: '@skyring/scylladown'
storage__path: 'skyring-1'
storage__collectionPoints: 'scylla-1:9042,scylla-2:9042,scylla-3:9042' (this line will be replaced with the headless service for scylla in Kubernetes`
Is there something obviously wrong I am doing here? Scylladb is supported oob here right?

File: kubernetes/skyring/templates/statefulset.yaml [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+              value: "*"
+            {{- end}}
+            - name: channel__host
+              valueFrom:
+                fieldRef:
+                  fieldPath: metadata.name
+            - name: nats__hosts
+              value: '{{ .Release.Name }}-nats-client:4222'
+            - name: seeds
+              value: '{{ .Values.nameOverride}}-seed:3455'
```
[the remark this answers is not in the document]
Comment (Iaskebba) [reply kubernetes/skyring/templates/statefulset.yaml]: @esatterwhite Yes absolutely makes sense. This was just a quick mock docker compose I made the error was in contactPoints. I need a quay image that has the mongo dependency and a quay image that has the scylladb dependency then I can make a pr that has leveldown w/ persistent storage, mongo, and scylladb support. 

I tested all three of them fairly heavily using my own images and they are really resilient to failures. 

There are a few bugs however but these are fine:

1. When using mongo for persistency if the Mongo Primary dies it can not recover as fast as needed - if timers try to clear or create while Mongo primary is down the node will crash

2. If a node dies while handling a large amount of timers a race condition will occur between timers firing and timers being cleared out as it passes off the timers to other nodes during shutdown. Since Kubernetes clusters can be run across multiple datacenters and nodes I propose the option to not pass off the timers at all during shutdown, I made these changes on a fork and tested it and this fits the Kubernetes structure much better. If a node dies that same pod can appear on another instance with the same Persistent Volume and recovery will bring back all the timers quickly and accurately. Furthermore, if a node crashes in the first place it often was unable to pass off timers anyways, so I felt that just standardizing on crash-like scenarios for Kubernetes always preserved the timers and their accuracy as pods can spin up in under 1-2 seconds after a crash for little to no downtime. 

3. Very rarely when a node dies - crash or regular shutdown, timers will RARELY be repeated. In a scenario of creating 1,000 timers a minute for 10 seconds every second and killing a node in the middle of this roughly 1 timer would fire twice every second trial. This is due to a race condition in the timer firing and clearing out persistent storage, I can create a pr for this as I identified where exactly this is.

So to summarize: When nodes don't pass off timers and just use the leveldown recovery the timers are incredibly resilient and due to Kubernetes speed of being able to spin up the same pods after a crash in a different datacenter these timers will always fire at accurate times.

[the remark this answers is not in the document]
Comment (esatterwhite) [reply kubernetes/skyring/templates/statefulset.yaml]: thanks for the detailed report. 
It sounds mostly like some work needs to go into the rebalance operations to make sure timers aren't executed when they are being moved. It sounds like the meat of the problem.

And you're right. mongo probably isn't the best choice. I originally set up mongo as an example - that remote storage is possible ( and people love mongo ).

a local leveldb is always going to be the best option.
scylla is going to be a pretty close 2nd if you really want to off load storage.

I'll be working on #27 for a while but the rebalance stuff is next on the list.

[the remark this answers is not in the document]
Comment (esatterwhite) [reply kubernetes/skyring/templates/statefulset.yaml]: you could add the two additional package to the docker image

https://github.com/esatterwhite/skyring/blob/master/Dockerfile#L18

```
npm install && npm install mongodown @skyring/scylladown --no-save ...
```
